### PR TITLE
🐛 [tool] Installs the APK regardless of version

### LINKED ISSUE #160402: [tool] Unable to install APK through `flutter run` when uninstalled once and installed a previous version
### Steps to reproduce

1. Set the `pubspec.yaml` version to `1.0.0+1`.
2. `flutter build apk --release && adb install build/app/outputs/flutter-apk/app-release.apk`
3. Run the app on the device, and ensure it is `v1.0.0+1`.
4. Set the `pubspec.yaml` version to `1.0.1+2`.
5. `flutter run -d the_android_device_id`
6. Ensure it is `v1.0.1+2`.
7. Uninstall the current `v1.0.1+2`, then reinstall the previous version through `adb install build/app/outputs/flutter-apk/app-release.apk`.
8. Run the app on the device, and ensure it is `v1.0.0+1`.
9. `flutter run -v -d the_android_device`
10. The log shows `Latest build already installed` and launched the app, but the latest APK was not installed.

### Expected results

`flutter run` should installs the latest APK.install

### Actual results

When a temporary file of a previously installed version exists, the tool will recognize that the latest version is installed.

### Code sample

N/A

### Screenshots or Video

N/A

### Logs

<details open><summary>Logs</summary>

```console
[   +2 ms] Stopping app 'app-debug.apk' on ONEPLUS A6010.
[        ] executing: /Users/alex/dev/android-sdk/platform-tools/adb -s f8e67ce3 shell am force-stop com.example.test
[  +95 ms] executing: /Users/alex/dev/android-sdk/platform-tools/adb -s f8e67ce3 shell pm list packages com.example.test
[ +109 ms] package:com.example.test
[        ] executing: /Users/alex/dev/android-sdk/platform-tools/adb -s f8e67ce3 shell cat /data/local/tmp/sky.com.example.test.sha1
[  +61 ms] a46ba751dfd13d0c0c4e153c8f8c52cc00b2a627
[        ] Latest build already installed.
[        ] executing: /Users/alex/dev/android-sdk/platform-tools/adb -s f8e67ce3 shell -x logcat -v time -t 1
[ +242 ms] --------- beginning of main
           12-17 09:53:38.172 I/cmd     (27942): oneway function results will be dropped but finished with status OK and parcel size 4
[   +3 ms] executing: /Users/alex/dev/android-sdk/platform-tools/adb -s f8e67ce3 shell am start -a android.intent.action.MAIN -c android.intent.category.LAUNCHER -f 0x20000000 --ez enable-dart-profiling true --ez enable-checked-mode true --ez verify-entry-points true --ez start-paused true com.example.test/com.example.test.MainActivity
[  +88 ms] Starting: Intent { act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x20000000 cmp=com.example.test/.MainActivity (has extras) }
[        ] Waiting for VM Service port to be available...

```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[!] Flutter (Channel [user-branch], 3.27.0-1.0.pre.753, on macOS 14.7.1 23H222 darwin-arm64, locale zh-Hans-CN)
```

</details>

### PR DESCRIPTION
**Pull request by AlexV525**
Resolves https://github.com/flutter/flutter/issues/160402

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by andrewkolos** [comment]
Adding @flutter/android-reviewers in case someone with more Android expertise can think of an easy way to fix this caching behavior rather than remove it. I suspect this change would add a few seconds (possibly more for wireless debugging users) to successive runs of `flutter run`. I think this is a fair tradeoff.

**Comment by reidbaker** [comment]
I have not reviewed the code and the tree is closed today because we are merging the engine and framework repos. 

Directionaly always installing an apk seems reasonable to me.

**Comment by AlexV525** [comment]
@andrewkolos I saw you removed the `isLatestBuildInstalled` from the code. That method still has its value IMO, you can still verify if the current build was installed through tests (although it might not be like the issue's case)

**Comment by andrewkolos** [comment]
> @andrewkolos I saw you removed the `isLatestBuildInstalled` from the code. That method still has its value IMO, you can still verify if the current build was installed through tests (although it might not be like the issue's case)

The implementation in `AndroidDevice` is bugged (hence this PR/issue), and the rest of the implementations either return `true` or `false` or throw an `UnimplementedError`. It doesn't appear to me that calling this from tests would be useful for verifying anything

File: packages/flutter_tools/test/general.shard/android/android_install_test.dart [hunk 1]
```diff
@@ -82,7 +82,7 @@ void main() {
       processManager: processManager,
     );
 
-    expect(await androidDevice.installApp(androidApk), false);
+    expect(await androidDevice.installApp(androidApk), isFalse);
```
Comment (andrewkolos) [inline packages/flutter_tools/test/general.shard/android/android_install_test.dart:85]: In my opinion, this change is not worth generating a diff. `isTrue`/`isFalse` don't generate better failure output, and I don't find this line of code to be significantly more readable than the original[^1].

See https://github.com/dart-lang/test/issues/2358

[^1]: If it is decided to be more readable, then we would have to consider using the `equals` matcher in place of bare values throughout the rest of the tool code.

Comment (AlexV525) [reply packages/flutter_tools/test/general.shard/android/android_install_test.dart]: TIL

**Review by andrewkolos — APPROVED** [review]
LGTM. However, I plan on filing a follow-up issue to see if caching can be restored—I don't feel great about several seconds being added to many `flutter run` invocations.

**Pull request merged**